aws/resource_aws_lb_listener_rule.go

```diff
@@ -283,18 +284,157 @@ func resourceAwsLbbListenerRule() *schema.Resource {
 			"condition": {
```

Comment:
There is also a max number of 5 conditions as per the UI

Reply:
This is difficult to catch. We can put a `MaxItems: 5` on the condition but users will still be hitting errors when trying to apply.

It's not just 5 condition blocks but 5 condition items. So you can have 5 totally different conditions with a single value or 5 values within a single host-header block.

It's also a current limitation. It may be changed in future?

Reply:
We can perhaps validate that using the `CustomizeDiff` function with the right check, so that we can still plan it and make it fail?

Reply:
I've been trying to get useful action out of `CustomizeDiff` and not getting very far. Lack of documentation isn't helping.

For some reason the new state in the diffset has all the list values set to nil instead of the new values. This was also causing the `lbListenerRuleConditionSetHash` function to explode too.

The change output displayed:
```
  ~ aws_lb_listener_rule.test
      condition.1014337362.field:                   "" => "path-pattern"
      condition.1014337362.path_pattern.#:          "0" => "1"
      condition.1014337362.path_pattern.0.values.#: "0" => "6"
      condition.1014337362.path_pattern.0.values.0: "" => "1"
      condition.1014337362.path_pattern.0.values.1: "" => "2"
      condition.1014337362.path_pattern.0.values.2: "" => "3"
      condition.1014337362.path_pattern.0.values.3: "" => "4"
      condition.1014337362.path_pattern.0.values.4: "" => "5"
      condition.1014337362.path_pattern.0.values.5: "" => "6"
      condition.1014337362.values.#:                "" => <computed>
      condition.3525442329.field:                   "path-pattern" => ""
```

Added a CustomizeDiff to try and see how these data structures work:
```go
        CustomizeDiff: customdiff.All(
            func(diff *schema.ResourceDiff, v interface{}) error {
                if diff.HasChange("condition") {
                    o, n := diff.GetChange("condition")
                    log.Printf("DDD o %T: %v", o, o)
                    log.Printf("DDD n %T, %v", n, n)
                }
                return nil
            },
        ),
```
Output:
```
...DDD o *schema.Set: &{0x3ddc210 map[3525442329:map[field:path-pattern host_header:[] http_header:[] http_request_method:[] path_pattern:[map[values:[1 2 3 4 5]]] query_string:[] source_ip:[] values:[1 2 3 4 5]]] {{0 0} 1}}
...DDD n *schema.Set, &{0x3ddc210 map[22963242:map[field:path-pattern host_header:[] http_header:[] http_request_method:[] path_pattern:[<nil>] query_string:[] source_ip:[] values:[]]] {{0 0} 1}}
```
Surely the second output should have `path_pattern:[1 2 3 4 5 6]` not `path_pattern:[<nil>]`? It's useless for our purposes in this state.

Anybody have any clues what might be wrong here?

Reply:
Looking at the source code it seems is doing nothing suspicious ( https://github.com/hashicorp/terraform/blob/master/helper/schema/resource_data.go#L489 ) although I would be curious to know what is the temporary output of the JSON that TF brings. Maybe that one is already corrupted and has nothing to do with `CustomDiff`.

Is it possible to attach it here so we could further analyze it? Thank you in advance.

Reply:
Hi @julianxhokaxhiu

Please see this gist: https://gist.github.com/dpiddockcmp/bd45caa17caa445c46cf9325da649789

Trace for this is still showing `nil` as the contents of the list:
```
... DDD o *schema.Set: &{0x421e2a0 map[3891208463:map[field:path-pattern host_header:[] http_header:[] http_request_method:[] path_pattern:[map[values:[1]]] query_string:[] source_ip:[] values:[1]]] {{0 0} 1}}
... DDD n *schema.Set, &{0x421e2a0 map[22963242:map[field:path-pattern host_header:[] http_header:[] http_request_method:[] path_pattern:[<nil>] query_string:[] source_ip:[] values:[]]] {{0 0} 1}}
```

Reply:
Does it still make sense to review this or do you want to continue like you mentioned in your last comment?

Reply:
Would be nice to know what I've done wrong if anything, but there's no point expanding too much effort here if I'm going to make more changes.

Reply:
I've thought about this some and done some playing around after changing to schema.Set. Setting list items to `<nil>` actually makes sense for new diffs. It's perfectly reasonable for the value to not exist yet. Or if one value in a long list is unknown, do you set that item individually as nil, or not include it in the list thus messing up the other index values? For example in the very common usecase of referencing another resource that hasn't been created yet.

schema.Sets muddy the water a little bit. New diff Sets contain values if they are known, or nothing if the value is unknown. This gives the situation of a Set having length Zero because the referenced resource has not been created yet at diff time.

Simple example:
```
resource "aws_eip" "this" {
  vpc = true
}

resource "aws_lb_listener_rule" "rule" {
  condition {
    source_ip = ["${aws_eip.this.public_ip}/32"]
  }
}
```

These review requests to add Diff time validation of number of rules or error handling is impossible to achieve with Terraform's current design.